plugins/janus_sipre.c

```diff
@@ -0,0 +1,3832 @@
+				if(strlen(custom_headers))
+					data = g_strdup(custom_headers);
+				/* TODO Any way to specify TTL in sipreg_register? */
+				char ttl_text[20];
+				g_snprintf(ttl_text, sizeof(ttl_text), "%d", ttl);
+				/* We enqueue this REGISTER attempt, to be sure it's done in the re_main loop thread
+				 * FIXME Maybe passing a key to the session is better than passing the session object
+				 * itself? it may be gone when it gets handled... won't be an issue with the
+				 * reference counter branch but needs to be taken into account until then */
+				mqueue_push(mq, janus_sipre_mqueue_event_do_register,
+					janus_sipre_mqueue_payload_create(session, NULL, 0, data));
+				result = json_object();
+				json_object_set_new(result, "event", json_string("registering"));
+			} else {
+				JANUS_LOG(LOG_VERB, "Not sending a SIPre REGISTER: either send_register was set to false or guest mode was enabled\n");
+				session->account.registration_status = janus_sipre_registration_status_disabled;
+				result = json_object();
+				json_object_set_new(result, "event", json_string("registered"));
+				json_object_set_new(result, "username", json_string(session->account.username));
+				json_object_set_new(result, "register_sent", json_false());
+				/* Also notify event handlers */
+				if(notify_events && gateway->events_is_enabled()) {
+					json_t *info = json_object();
+					json_object_set_new(info, "event", json_string("registered"));
+					json_object_set_new(info, "identity", json_string(session->account.identity));
+					json_object_set_new(info, "type", json_string("guest"));
+					gateway->notify_event(&janus_sipre_plugin, session->handle, info);
+				}
+			}
+		} else if(!strcasecmp(request_text, "call")) {
+			/* Call another peer */
+			if(session->account.registration_status < janus_sipre_registration_status_registered) {
+				JANUS_LOG(LOG_ERR, "Wrong state (register first)\n");
+				error_code = JANUS_SIPRE_ERROR_WRONG_STATE;
+				g_snprintf(error_cause, 512, "Wrong state (register first)");
+				goto error;
+			}
+			if(session->status >= janus_sipre_call_status_inviting) {
+				JANUS_LOG(LOG_ERR, "Wrong state (already in a call? status=%s)\n", janus_sipre_call_status_string(session->status));
+				error_code = JANUS_SIPRE_ERROR_WRONG_STATE;
+				g_snprintf(error_cause, 512, "Wrong state (already in a call? status=%s)", janus_sipre_call_status_string(session->status));
+				goto error;
+			}
+			JANUS_VALIDATE_JSON_OBJECT(root, call_parameters,
+				error_code, error_cause, TRUE,
+				JANUS_SIPRE_ERROR_MISSING_ELEMENT, JANUS_SIPRE_ERROR_INVALID_ELEMENT);
+			if(error_code != 0)
+				goto error;
+			json_t *uri = json_object_get(root, "uri");
+			/* Check if we need to ACK manually (e.g., for the Record-Route hack) */
+			json_t *autoack = json_object_get(root, "autoack");
+			gboolean do_autoack = autoack ? json_is_true(autoack) : TRUE;
+			/* Check if the INVITE needs to be enriched with custom headers */
+			char custom_h
```

Comment:
Thanks for the detailed analysis! I'm glad someone is looking into this, as I'd really like to merge this plugin soon (currently working on some other features it still lacks from the other SIP plugin), but without proper feedback it would be hard to see if this is actually in a usable state or not.

As to `dnsc`, we should probably check how baresip creates and handles it.